sql: fix DROP COLUMN bug which ignores containing indexes

otan opened the pull request:
AllNonDropIndexes returns an array of pointers to index descriptors.
After deleting one of the indexes, the array of pointers point
to a different element. This ends up accidentally missing some
indexes that require deletion.

Resolves #54196 (introduced in #51661).

Release note (bug fix): Fixed a bug introduced in earlier v20.2 versions
where when attempting to drop a column which is referenced by multiple
indexes fails to drop all relevant indexes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54228)
<!-- Reviewable:end -->

mgartner reviewed (changes requested):
By `GetAllNonDropIndexes` do you mean `AllNonDropIndexes`?

Can you explain more in the commit message what is causing indexes to be missed in the delete? It's not immediately obvious how my change created this problem. Or was the bug already there, but hidden by the strict dropping rules before my change?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-:-MGudJMl2IUtvlJ_hsme:b66agjb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @otan)

___
*[pkg/sql/alter_table.go, line 406 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-MGufGfp9lEhWWuLONXO:-MGufGfp9lEhWWuLONXP:b-z3nk8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c060249be2d6519ce2c3555c64363e186855624b/pkg/sql/alter_table.go#L406)):*
> ```Go
> 					"column %q is referenced by the primary key", colToDrop.Name)
> 			}
> 			var idxesToDelete []descpb.IndexDescriptor
> ```

Could we store just a list of `idx.Name` since that's the only thing `dropIndexesByName` needs?
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-MGudnAXCGaGhJe09j56:-MGudnAXCGaGhJe09j57:b3xq0ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/c060249be2d6519ce2c3555c64363e186855624b/pkg/sql/logictest/testdata/logic_test/alter_table#L1433)):*
> ```
> 		STORING(col109_2, col109_3, col109_6, col109_16),
> 	INDEX (col109_12 ASC, col109_4 DESC)
> ); ALTER TABLE regression_54196 DROP COLUMN col109_15
> ```

Can this test case be simplified?

I think you should add a `SHOW INDEXES FROM regression_54196` to assert the correct indexes have been dropped.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
AllNonDropIndexes returns an array of pointers to index descriptors.
After deleting one of the indexes, the array of pointers point
to a different position. This ends up accidentally missing some
indexes that require deletion.

updated commit message ^

i think the index drops when columns drop were added during your change, hence starting the problem
beforehand, we failed to allow drop column when it was referenced by an index, hehe.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-:-MGulZyiADqGg5fUTpjt:b907vxu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/alter_table.go, line 406 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-MGufGfp9lEhWWuLONXO:-MGukm5t2kHUYe8Y997k:bazyd9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c060249be2d6519ce2c3555c64363e186855624b/pkg/sql/alter_table.go#L406)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Could we store just a list of `idx.Name` since that's the only thing `dropIndexesByName` needs?
</blockquote></details>

Yeah I can do that.
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-MGudnAXCGaGhJe09j56:-MGukiT_ClWinjMizkNl:b32x1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c060249be2d6519ce2c3555c64363e186855624b/pkg/sql/logictest/testdata/logic_test/alter_table#L1433)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can this test case be simplified?

I think you should add a `SHOW INDEXES FROM regression_54196` to assert the correct indexes have been dropped.
</blockquote></details>

ya


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix DROP COLUMN bug which ignores containing indexes

AllNonDropIndexes returns an array of pointers to index descriptors.
After deleting one of the indexes, the array of pointers point
to a different element. This ends up accidentally missing some
indexes that require deletion.

Release note (bug fix): Fixed a bug introduced in earlier v20.2 versions
where when attempting to drop a column which is referenced by multiple
indexes fails to drop all relevant indexes.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from @cockroachdb/sql-schema 

Looking forward to pulling these mutations underneath the mutable descriptor and generally just get rid of mucking around with the protos directly. It's not clear that that alone would solve anything but it would (hopefully) lead to unit testing that *might* have caught this sort of thing. I have a dream of doing property based testing for sequences of mutations and their impact on retreival methods but that feels not supremely close. 

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-:-MGurHK9563O2GYMckIi:bl17en0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Yikes... The index descriptors are stored as `[]IndexDescriptor` in `tabledesc.Immutable`. When `dropIndexByName` is called, the all the indexes after the dropped index are shifted to the left by one, but the pointers from `AllNonDropIndexes` remain the same, so an index is skipped.

That's nasty. I suppose when I wrote it I was not aware that `dropIndexByName` mutated the underlying slice of IndexDescriptors. LMK if you have ideas on how to defend against something like this in the future.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for tracking this down.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54228#-:-MGurvBr2SO0zYVhQQeW:b3t2hzy)* status: :shipit: complete! 2 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
Yeah it's not great how easy this mistake can come about. I think schema's got ideas, I chanced upon this one :P 





<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajwerner,mgartner

backport will come
